### PR TITLE
Allow ReflectionPad to accept 0-dim batch sizes.

### PR DESCRIPTION
**Pull request by v0dro**
Allows ReflectionPad 1D and 2D to accept 0-dim batch sizes. 

Related to issues:

* https://github.com/pytorch/pytorch/issues/38115
* https://github.com/pytorch/pytorch/issues/12013

### CODE REVIEW

**Comment by mruberry** [comment]
CUDA test failures are real. Do you know what the issue is?

**Comment by v0dro** [comment]
@mruberry problems have been resolved. Could you please review?

File: test/test_nn.py [hunk 1]
```diff
@@ -9382,6 +9382,12 @@ def test_GroupNorm_empty(self, device):
             with torch.backends.cudnn.flags(enabled=False):
                 self._test_module_empty_input(mod, inp)
 
+    def test_ReflectionPad_empty(self, device):
+        for mod, inp in [
+                (torch.nn.ReflectionPad1d(2), torch.randn(0, 3, 10, device=device)),
```
Comment (mruberry) [inline test/test_nn.py:9387]: I don't think you're covering all the cases? Don't you need to check ReflectionPad1D with a 2D and 3D input with a zero batch size plus ReflectionPad2D with a 3D and 4D input with a zero batch size?

You should also add test cases that fail, where the non-batch dimension is zero and you assert you hit the appropriate error (`self.assertRaisesRegex(...)`), and a check for backward when the batch dim is zero.

**Comment by mruberry** [comment]
> @mruberry problems have been resolved. Could you please review?

Looks pretty good and clean, but you need to add more test cases to validate all the new behavior.

File: test/test_nn.py (hunk 1 above)
Comment (v0dro) [reply test/test_nn.py:9387]: Ah yes good point. Will add them thank you.

Comment (v0dro) [reply test/test_nn.py:9387]: The checks for 2D/3D input are done in the python code in `functional.py` and the `ReflectionPad1D` does not get called unless the input is specifically `dim=3`. Same goes for 2D reflection pad (with `dim=4`). Therefore such tests are not needed.

However I have added tests for non-batch dimension being zero. The backward is tested in the `_test_module_empty_input`.

**Comment by v0dro** [comment]
@mruberry can you please review this PR again?

**Review by mruberry — APPROVED** [review]
Thanks for the follow-up, v0dro! I added a decorator to the test that should resolve the XLA build failure. If the tests pass this should be good to go!

**Comment by mruberry** [comment]
This was successfully merged. Thanks v0dro!
